cl/citations/annotate_citations.py

```diff
@@ -0,0 +1,107 @@
+from typing import List
+
+from eyecite import annotate, clean_text
+from eyecite.models import CitationBase, NonopinionCitation
+
+from cl.search.models import Opinion
+
+
+def get_and_clean_opinion_text(opinion: Opinion) -> None:
+    """Memoize useful versions of an opinion's text as additional properties
+    on the Opinion object. This should be done before performing citation
+    extraction and annotation on an opinion.
+
+    :param opinion: The Opinion whose text should be parsed
+    """
+    if opinion.html_anon_2020:
+        opinion.source_text = opinion.html_anon_2020
+        opinion.cleaned_text = clean_text(
+            opinion.html_anon_2020, ["html", "all_whitespace"]
+        )
+        opinion.source_is_html = True
+    elif opinion.html_columbia:
+        opinion.source_text = opinion.html_columbia
+        opinion.cleaned_text = clean_text(
+            opinion.html_columbia, ["html", "all_whitespace"]
+        )
+        opinion.source_is_html = True
+    elif opinion.html_lawbox:
+        opinion.source_text = opinion.html_lawbox
+        opinion.cleaned_text = clean_text(
+            opinion.html_lawbox, ["html", "all_whitespace"]
+        )
+        opinion.source_is_html = True
+    elif opinion.html:
+        opinion.source_text = opinion.html
+        opinion.cleaned_text = clean_text(
+            opinion.html, ["html", "all_whitespace"]
+        )
+        opinion.source_is_html = True
+    elif opinion.plain_text:
+        opinion.source_text = opinion.plain_text
+        opinion.cleaned_text = clean_text(
+            opinion.plain_text, ["all_whitespace"]
+        )
+        opinion.source_is_html = False
+
+
+def generate_annotations(
+    citations: List[CitationBase],
+) -> List[List]:
```

Comment:
I want the return signature here to be `List[List[Union[Tuple[int, int], str]]]`, i.e., a list of lists where the inner list can contain either strings or tuples of ints. However, this does not seem to work, and I'm not sure what I'm doing wrong.

Reply:
Eh, let's not force it. If the above passes, I'm sure we can dial it in further later when we're better at mypy.